src/sctools/test/test_count.py

```diff
@@ -1,44 +1,578 @@
+            count_matrix[:, i_gene] = non_zero_count_matrix[:, i]
+        return count_matrix
+
+    def _generate_primary_alignment_record_bundle(self, count_matrix: np.ndarray) \
+            -> Tuple[Set[AlignmentRecordTags], List[str], List[str]]:
+        alignments: Set[AlignmentRecordTags] = set()
+        used_cell_barcodes: Set[str] = set()
+
+        row_index: List[str] = []
+        col_index = self.all_gene_names
+
+        for i_cell in range(self.num_cells):
+            # generate a unique cell barcode
+            while True:
+                cell_barcode = self._generate_random_cell_barcode()
+                if cell_barcode not in used_cell_barcodes:
+                    break
+            row_index.append(cell_barcode)
+
+            for i_gene in self.to_be_used_gene_indices:
+                for i_molecule in range(count_matrix[i_cell, i_gene]):
+                    # generate a unique alignment tag
+                    unique_alignment_tag = self._generate_unique_random_alignment_tag(
+                        alignments,
+                        gene_name=self.all_gene_names[i_gene],
+                        cell_barcode=cell_barcode)
+                    alignments.add(unique_alignment_tag)
+
+        return alignments, row_index, col_index
+
+    def _generate_unique_random_alignment_tag(self,
+                                              existing_alignment_tags: Set[AlignmentRecordTags],
+                                              gene_name: str,
+                                              cell_barcode: Optional[str] = None,
+                                              molecule_barcode: Optional[str] = None) -> AlignmentRecordTags:
+        assert gene_name in self.to_be_used_gene_names, f"{gene_name} is not an allowed gene for generating " \
+                                                        f"synthetic data"
+        while True:
+            alignment = AlignmentRecordTags(
+                cell_barcode=cell_barcode if cell_barcode else self._generate_random_cell_barcode(),
+                molecule_barcode=molecule_barcode if molecule_barcode else self._generate_random_molecule_barcode(),
+                gene_name=gene_name)
+            if alignment not in existing_alignment_tags:
+                return alignment
+
+    def _generate_duplicate_alignment_tags(self,
+                                           num_duplicates: int,
+                                           primary_alignments_list: List[AlignmentRecordTags])\
+            -> List[AlignmentRecordTags]:
+        return self.rng.choice(primary_alignments_list, size=num_duplicates).tolist()
+
+    def _generate_incomplete_alignment_tags(self, num_missing_some_tags: int) -> List[AlignmentRecordTags]:
+        incomplete_alignment_tags_list: List[AlignmentRecordTags] = list()
+        for _ in range(num_missing_some_tags):
+            tag_mask = self.rng.randint(low=0, high=7)
```

Comment:
Nice, I like this way of masking. Very efficient!

Reply:
:)